### PR TITLE
Add panel fragments to read

### LINKED ISSUE #43: SAP2000_Toolkit: ReadPanel() to get additional properties as Fragment
#### Description:

Create Fragment for panels to capture:

- [ ] Auto Mesh
- [ ] Edge Constraint
- [ ] Material Overwrite
- [ ] Notional Size
- [ ] Offsets
- [ ] Transformation Matrix

### PR DESCRIPTION
**Pull request by JosefTaylor**
### NOTE: Depends on 

https://github.com/BHoM/Revit_Toolkit/pull/1130

   
### Issues addressed by this PR

Closes #43 

### Test files

https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/SAP2000_Toolkit/%23241-Panel-Fragments?csf=1&web=1&e=qmCvdW

### CODE REVIEW

**Comment by JosefTaylor** [comment]
@BHoMBot check installer

**Comment by JosefTaylor** [comment]
@BHoMBot check copyright

**Comment by JosefTaylor** [comment]
@BHoMBot check copyright-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot fix copyright headers ref. 3557557496

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot fix project file ref. 4305558738

**Comment by JosefTaylor** [comment]
@BHoMBot check project-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check documentation-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot fix documentation punctuation ref. 4314388387

**Comment by JosefTaylor** [comment]
@BHoMBot check documentation-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check documentation-compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check versioning

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check core

**Comment by JosefTaylor** [comment]
@BHoMBot check serialisation

**Comment by JosefTaylor** [comment]
@BHoMBot check installer

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

File: SAP2000_oM/Fragments/IPanelAutoMesh.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    [Description("Base interface for panel auto mesh settings in SAP2000. Contains the type of auto mesh and the properties common to all types.")]
+    [Unique]
+    public interface IPanelAutoMesh : IFragment
+    {
+        [Description("If this item is True, and if both points along an edge of the original area " + "object have the same local axes, the program makes the local axes for added points " + "along the edge the same as the edge end points.")]
+        bool LocalAxesOnEdge
+        {
+            get;
+            set;
+        }
```
Comment (IsakNaslundBh) [inline SAP2000_oM/Fragments/IPanelAutoMesh.cs:37]: Not sure why all of these are on different lines?

Can we change all of them to be single line properties as the general convention. Especially when default values are assigned keeping them on multiple lines makes it really quite hard to read.

```suggestion
        bool LocalAxesOnEdge { get; set; }
```

**Review by IsakNaslundBh — DISMISSED** [review]
Think the oM properties needs adjusting in terms of formatting, apart from that, the code looks fine. Not able to test as I do not have SAP2000 though

**Comment by JosefTaylor** [comment]
@enarhi @mchaf would you do the honor of testing?

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
Fixed various reference issues lingering from the rebase, @IsakNaslundBh @enarhi @mchaf would you mind looking over?

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

File: SAP2000_Engine/Query/Panel.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public static partial class Query
+    {
+        [Description("Returns the SAP2000 PanelAutoMesh settings for a panel. You can also use the method FindFragment() with the type IPanelAutoMesh as an argument.")]
+        [Input("panel", "A panel which was either pulled from SAP2000 or which has had SAP2000 settings added.")]
+        [Output("PanelAutoMesh", "A fragment containing SAP2000 PanelAutoMesh settings.")]
+        public static IPanelAutoMesh PanelAutoMesh(this Panel panel)
+        {
+            List<IFragment> fragments = panel?.GetAllFragments(typeof(IPanelAutoMesh));
+            if (fragments.Count > 1) Compute.RecordWarning($"the panel {panel.Name} has more than one PanelOffset defined, which is not allowed. Only the first has been returned. Use GetAllFragments() to extract others.");
+            return fragments.Select(x => x as IPanelAutoMesh).FirstOrDefault();
```
Comment (IsakNaslundBh) [inline SAP2000_Engine/Query/Panel.cs:46]: This method will fail the nullcheck. Something like this should sort it I think. (if the panel is null, then the list of fragments will also be null from the `?` )

```suggestion
            List<IFragment> fragments = panel?.GetAllFragments(typeof(IPanelAutoMesh));
            
            if(fragments == null || fragments.Count == 0)
               retrun null;
            
            if (fragments.Count > 1) Compute.RecordWarning($"the panel {panel.Name} has more than one PanelOffset defined, which is not allowed. Only the first has been returned. Use GetAllFragments() to extract others.");
            return fragments.Select(x => x as IPanelAutoMesh).FirstOrDefault();
```

File: SAP2000_Engine/Query/Panel.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        [Description("Returns the SAP2000 PanelOffset settings for a panel. You can also use the method FindFragment() with the type IPanelOffset as an argument.")]
+        [Input("panel", "A panel which was either pulled from SAP2000 or which has had SAP2000 settings added.")]
+        [Output("PanelOffset", "A fragment containing SAP2000 PanelOffset settings.")]
+        public static IPanelOffset PanelOffset(this Panel panel)
+        {
+            List<IFragment> fragments = panel?.GetAllFragments(typeof(IPanelOffset));
+            if (fragments.Count > 1) Compute.RecordWarning($"the panel {panel.Name} has more than one PanelOffset defined, which is not allowed. Only the first has been returned. Use GetAllFragments() to extract others.");
+            return fragments.Select(x => x as IPanelOffset).FirstOrDefault();
```
Comment (IsakNaslundBh) [inline SAP2000_Engine/Query/Panel.cs:64]: As above

```suggestion
            List<IFragment> fragments = panel?.GetAllFragments(typeof(IPanelOffset));
                        
            if(fragments == null || fragments.Count == 0)
               retrun null;
            
            if (fragments.Count > 1) Compute.RecordWarning($"the panel {panel.Name} has more than one PanelOffset defined, which is not allowed. Only the first has been returned. Use GetAllFragments() to extract others.");
            return fragments.Select(x => x as IPanelOffset).FirstOrDefault();
```

**Review by IsakNaslundBh — COMMENTED** [review]
Commetns on methods probably cause to the null handling failures

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by JosefTaylor** [comment]
@BHoMBot check required

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance
@BHoMBot check null-handling

**Comment by IsakNaslundBh** [comment]
@JosefTaylor I had to rebase and force push this PR again after #253 was merged. as usual, please delete the branch locally before pulling down changes again! 🛩️

**Comment by JosefTaylor** [comment]
@BHoMBot check installer

**Comment by JosefTaylor** [comment]
@BHoMBot check versioning

**Comment by JosefTaylor** [comment]
@BHoMBot check compliance

**Comment by JosefTaylor** [comment]
@BHoMBot check null-handling

**Comment by JosefTaylor** [comment]
@BHoMBot check core
@BHoMBot check serialization

**Comment by JosefTaylor** [comment]
@BHoMBot check serialisation

**Comment by enarhi** [comment]
@BHoMBot check required

**Review by enarhi — APPROVED** [review]
LGTM, tested using the provided test script and fragments were assigned properly, pushed to SAP, and pulled as expected.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance
@BHoMBot check ready-to-merge

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am authorising dispensation to be granted on check ref. 10538727097

**Comment by FraserGreenroyd** [comment]
Although there are new files, I have dispensated the check in favour of doing a smaller, newer, surgical PR tomorrow to update copyright headers again as this PR has been open a while and already rebased once so we don't want to keep it open too much longer.

**Pull request merged**